### PR TITLE
Feat(toc): Introduce

### PR DESCRIPTION
**Pull request by AugustinMauroy**
## Description

Migration of table of contents from nodejs.dev but using nextra TOC

## Related Issues

#5242 

### Check List

- [X] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [X] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [X] I have run `npx turbo format` to ensure the code follows the style guide.
- [X] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [X] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Comment by AugustinMauroy** [comment]
i will rebase/cherrypick to fix everything don't worry

**Comment by AugustinMauroy** [comment]
Rebase done !

File: components/Article/TableOfContents/index.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  id: string;
+  children?: TableOfContentsNode[];
+  parent?: TableOfContentsNode | null;
+};
+
+type TableOfContentsProps = {
+  tableOfContents: TableOfContentsNode[];
+};
+
+function parseTableOfContents(
```
Comment (ovflowd) [inline components/Article/TableOfContents/index.tsx]: I see that your logic is a little bit based on how Nextra Headings work, but your logic is super over complicated and far from performance efficient. In simple words, it's utter mess.

I think there's a lot of room for improvement here (See the simplicity of Nextra's TOC component (https://github.com/shuding/nextra/blob/main/packages/nextra-theme-docs/src/components/toc.tsx)) the logic itself is just a simple `array.map` and nothing more.

Pretty much it uses a clever approach of using class names for different depths. And these class names adjust the left-marging and the font-size. I think that's honestly a really good and simple approach.

**Review by ovflowd — CHANGES_REQUESTED** [review]
I appreciate the effort and work done here. But the current state of the PR is far from something we can ship in this repository.

I left some feedback, but I would rather not giving code examples ad I want you to think by yourself (this time) and be able to achieve something useful without the constant need of mentoring.

File: components/Article/TableOfContents/index.tsx (hunk 1 above)
Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: We can do that but. We will not have nesting `<ul>`.
Just to know the pasing was done on front-end or in SSR ?

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: We don't need `ul`.

> Just to know the pasing was done on front-end or in SSR ?

I'm not sure what you're saying.

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: > I'm not sure what you're saying.

I mean it's client component or server component ?

Comment (HinataKah0) [reply components/Article/TableOfContents/index.tsx]: @AugustinMauroy 

Hi Augustin,

I just take a look at both codes (yours and Nextra's). There are many ways to implement/code things. 😅 

Something simpler is usually better (coding together is about teamwork). For example: I understand Nextra's code (and how it works) in under 2 mins (it is simply an `items.map()` and each item has `id`, `value`, and `depth`).

But, it takes more time to understand this PR. 😆 
What you write is called "Iterative DFS" (let's Not discuss about this, I just wanted to show that it is something with a cool name). And not all people know about it.

So, I think it's better to keep things simpler. What do you think?
Maybe you can try to look into how Nextra's implement the ToC and try to understand it first. :)
Try to understand why things work... It will be useful for you...

Don't worry too much about Server Side vs Client Side for now.
Let's not worry about too many things, we can focus on one thing at a time.

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: I'm taking a look to css instead of this pice of incredible complexity

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: Hey, @AugustinMauroy, any update based on @HinataKah0's feedback? 👀

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: CSS part I have the solution. I'm looking on nextjs part. Because now we doesn't have nextra.

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: Why would you need Nextra here? Their Navigation component is agnostic, I'm not seeing the issue here.

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: The system was based on nextra toc.

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: What system are you talking about? We only mentioned that our ToC can be a simple Array loop (1-dimensional) and the rest be CSS, like Nextra.

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: Yeah but nextra give (in pagePros) an array with data but now we don't have this data so I have to create this data fetcher

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: > Yeah but nextra give (in pagePros) an array with data but now we don't have this data so I have to create this data fetcher

I have no idea what you're talking about. The component should be agnostic to whatever. Yes we do not use Nextra, but we can easily get the ToC by using plugins, such as https://github.com/remarkjs/remark-toc.

In the end, the data structure we pass to the TOC component should be agnostic and can be whatever we define it to be. It should pretty much contain a `title`, `level` and `id`; This should be the data structure of any TOC that comes from Markdown.

So you shouldn't worry about what Framework we use or not, the TOC component should be simple, have a generic data structure, since we can pretty much when calling the component map any data structure from the Framework/plugin we use to what the component supports.

**Comment by ovflowd** [comment]
@AugustinMauroy this PR https://github.com/nodejs/nodejs.org/pull/5467 introduces the feature that actually will support headings metadata so you can create the ToC.

This is the overall shape of the Headings:

```json
[
  {"depth": 1, "value": "Heading 1", "data": {"id": "custom-id"}},
  {"depth": 2, "value": "Heading 2"},
  {"depth": 3, "value": "Heading 3"},
]
```

And the type of `Heading` is also exported by `@vcarl/remark-headings`. So now you can adopt the approach we mentioned with the CSS and simple loops.

**Comment by AugustinMauroy** [comment]
@ovflowd What types of `data` ?

File: public/node-releases-data.json [hunk 2]
```diff
@@ -1 +1 @@
-[]
```
Comment (ovflowd) [inline public/node-releases-data.json:1]: Please do not commit these files.

Comment (ovflowd) [reply public/node-releases-data.json:1]: Run `npm ci` to install husky.

**Comment by ovflowd** [comment]
> @ovflowd What types of data ?

Could you rephrase your question? I have no idea what you're referring to/talking about.

File: __fixtures__/page.ts
Comment (ovflowd) [inline __fixtures__/page.ts:1]: Why is the name of this fixture `page`?

File: components/Article/TableOfContents/index.module.scss [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+
+  ul {
+    margin: var(--space-04) 0 0 var(--space-08);
+    padding: var(--space-08) 0 0 var(--space-24);
+
+    li {
+
+      // not use for now
```
Comment (ovflowd) [inline components/Article/TableOfContents/index.module.scss:22]: Please remove this commented code 😅

File: components/Article/TableOfContents/index.module.scss [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      //&.depth1 {
+      //
+      //}
+      &.depth2 {
+        list-style-type: square;
+        margin-left: var(--space-08);
+      }
```
Comment (ovflowd) [inline components/Article/TableOfContents/index.module.scss:29]: Empty line between each other statement... 👀

**Comment by bmuenzenmeyer** [comment]
@AugustinMauroy this PR and a couple other ones seem to be stagnating. Do you want help driving them toward completion?
